### DISCUSSION TITLE
Docker build from 2022-07-09 working :)

### DISCUSSION THREAD

**Discussion by accessiblepixel** [Aterfax/relax-intel-rmrr#15] [General]
Hey @Aterfax 

I've done my upgrade to Proxmox 7.2 (on my G8 HPE DL380e) and I've tried the kernel from https://github.com/Aterfax/relax-intel-rmrr/actions/runs/2639751612 and it works just fine passing through a PCIe Disk Controller, and I thought I'd just write a quick discussion thread and let you know :)

When I get some more time, I'll also try it on my G6s and see if it works on that too. 👍 

Thank you :)

Jessica

**Comment by Aterfax** [comment]
Ta, I'll look at adding this build to the releases soonish. (Need to schedule some downtime and give it a quick confirm on my own node.)

**Reply by accessiblepixel** [reply]
No worries, just thought I'd mention that it works, flawlessly I might add. Thanks! You're a star.

If you need me to test anything or whatnot, give me a shout and I'll do my best. Don't know much kernel code, but have compiled it a couple of times (not recently though haha)

I did have to do something a bit different to pass through some NVMe storage, but I think that was because of a firmware bug on them rather than anything to do with the kernel.

(had to modify my VM's config with `args: -set device.hostpci0.x-msix-relocation=bar2` )

With the docker build image, how does that run? Would you just download the docker image and run it and you get a kernel at the end then, or...? No rush on a reply if it's something complex.

Hope you had a good one.

Jessica

**Reply by Aterfax** [reply]
>(had to modify my VM's config with args: -set device.hostpci0.x-msix-relocation=bar2 )

Yes this certainly looks related to [some firmware / VFIO passthrough shenanigans](https://bugzilla.kernel.org/show_bug.cgi?id=202055). You may be able to avoid the need for this by flashing new firmware to your NVMe drive/s.


>With the docker build image, how does that run? Would you just download the docker image and run it and you get a kernel at the end then, or...? No rush on a reply if it's something complex.

Assuming you have docker installed on a machine you'd just run the command:

`
docker run --name relaxable-rmrr-proxmox-kernel-builder -v /mnt/scratch/proxmox-kernel-build-area/proxmox-kernel:/build/proxmox/proxmox-kernel -it aterfax/relaxable-rmrr-proxmox-kernel-builder:latest`

Where the [volume binding](https://docs.docker.com/storage/bind-mounts/) (-v bit) is set to the host machine folder of your choice, in the above case `/mnt/scratch/proxmox-kernel-build-area/proxmox-kernel`  Or the general example: `-v /host_machine/folder/path:/docker_container/folder/path`

Note that the Docker container path `/build/proxmox/proxmox-kernel` must remain the same as this is how the image is built to work.

That command will download the most recent Docker image for the builder and start building it. Give it 10 - 300 minutes depending on how fast your host machine and storage drives are and the debs will get generated in the host path you chose. Just make sure you give it a folder on a disk with ~40GB of spare space (the full kernel source is rather large and the intermediate build files will be too). 

(Don't expect the lower end of that timescale though, that's more what I see when running kernel compiles on a supercomputer node with some fairly high specs.)

More details on the docker process above can be found here: https://github.com/Aterfax/relax-intel-rmrr/blob/master/README.md#docker---build-packages-from-sources-intermediate

The Docker image itself is built from: https://github.com/Aterfax/relax-intel-rmrr/blob/master/Dockerfile after Git cloning the repo locally. You can also see that the [ENTRYPOINT ](https://docs.docker.com/engine/reference/builder/#entrypoint)[is the last line](https://github.com/Aterfax/relax-intel-rmrr/blob/master/Dockerfile#L36) and is set to run the [exact same command in the Github Action](https://github.com/Aterfax/relax-intel-rmrr/blob/master/.github/workflows/build-kernel-debs.yml#L21) since the Action will override the standard ENTRYPOINT.

This Docker build process is exactly the same [as the Github Action used to make the debs here on Github](https://github.com/Aterfax/relax-intel-rmrr/blob/master/.github/workflows/build-kernel-debs.yml), as the Github action is using the same docker image with a bit more Github Action tooling to save the important files (as artifacts). 

My action is also set to use [tagged Github Action runners](https://github.com/Aterfax/relax-intel-rmrr/blob/master/.github/workflows/build-kernel-debs.yml#L10) (tagged: pve-kernel) on virtual machines I have specifically configured to build these kernel debs This is mainly to do with using faster (SSD) storage, limiting the blast radius of any compromise with VLAN network separation, [retaining the build files between kernel builds](https://github.com/Aterfax/relax-intel-rmrr/blob/master/.github/workflows/build-kernel-debs.yml#L13) to speed things up and because the default Github Actions runners would take a very long time to finish compiling. (Plus I think they also lack the required storage space?)

This repo will also [automatically rebuild the docker image and upload it to dockerhub](https://github.com/Aterfax/relax-intel-rmrr/blob/master/.github/workflows/docker-publish.yml) when I make changes. The `aterfax/relaxable-rmrr-proxmox-kernel-builder:latest` is over here:  https://hub.docker.com/r/aterfax/relaxable-rmrr-proxmox-kernel-builder

If you are interested in the actions driving this, they are all stored in https://github.com/Aterfax/relax-intel-rmrr/blob/master/.github/workflows/

>If you need me to test anything or whatnot, give me a shout and I'll do my best. Don't know much kernel code, but have compiled it a couple of times (not recently though haha)

I think if anything making the readme / documentation around the process as concise and clear as possible would be great so more eyes on that and any suggestions / comments are welcome. I don't expect the build process / kernel compile parts to change a great deal or often so we're likely good on the more esoteric codey front for a while.

My plan is to update the kernel on my node over the weekend assuming I don't melt into a puddle (the UK being mid-heatwave by that time by the look of the forecast.)

**Reply by Aterfax** [reply]
I have also validated this build as functional. Thanks!

**Reply by Aterfax** [reply]
Release is now available.

**Reply by accessiblepixel** [reply]
Huzzah!

Thank you for the detailed information for using docker for builds, very appreciated.

Been a bit busy past couple of weeks, but the build I installed from this has been working flawlessly. I haven't updated to any later build yet, but I have a plan for testing newer builds without having to test them "trial by fire".

I've got a somewhat spare G6 here that I have a testing install of proxmox on, so I think that would make a perfect testing ground for testing passthrough and this kernel, so I can give more regular feedback.

I'll definitely have a look at the documentation to see if there's anything I can help clarify for readability and people less familiar with working round HPE's somewhat esoteric hardware :)

Hope you've had a good couple of weeks.

Kind regards,
Jessica.
